attribute-service-client/src/main/java/org/hypertrace/core/attribute/service/client/AttributeServiceCachedClient.java

```diff
@@ -30,102 +31,91 @@
-      String tenantId, String attributeScope, String attributeKey) {
+      @Nonnull RequestContext requestContext,
+      @Nonnull String attributeScope,
+      @Nonnull String attributeKey)
+      throws ExecutionException {
     try {
-      return Optional.ofNullable(cache.getUnchecked(tenantId))
+      return getTableForRequestContext(requestContext)
           .map(table -> table.get(attributeScope, attributeKey));
-    } catch (Exception e) {
-      if (LOG_LIMITER.tryAcquire()) {
-        log.error("No attribute available for scope {} and key {}", attributeScope, attributeKey);
-      }
+    } catch (NullPointerException e) {
+      log.debug("No attribute available for scope {} and key {}", attributeScope, attributeKey);
       return Optional.empty();
     }
   }
 
-  public Optional<AttributeMetadata> getById(String tenantId, String attributeId) {
+  public Optional<AttributeMetadata> getById(
+      @Nonnull RequestContext requestContext, @Nonnull String attributeId)
+      throws ExecutionException {
     try {
-      return Optional.ofNullable(cache.getUnchecked(tenantId))
-          .flatMap(
-              table ->
-                  Optional.ofNullable(scopeAndKeyLookup.getIfPresent(attributeId))
-                      .map(scopeAndKey -> table.get(scopeAndKey.getScope(), scopeAndKey.getKey())));
-    } catch (Exception e) {
-      if (LOG_LIMITER.tryAcquire()) {
-        log.error("No attribute available for id {}", attributeId);
-      }
+      return getTableForRequestContext(requestContext)
+          .map(
+              table -> {
+                AttributeScopeAndKey scopeAndKey = scopeAndKeyLookup.getIfPresent(attributeId);
+                return table.get(scopeAndKey.getScope(), scopeAndKey.getKey());
+              });
+    } catch (NullPointerException e) {
+      log.debug("No attribute available for id {}", attributeId);
       return Optional.empty();
     }
   }
 
-  public Optional<List<AttributeMetadata>> getAllInScope(String tenantId, String attributeScope) {
-    try {
-      return Optional.ofNullable(cache.getUnchecked(tenantId))
-          .map(table -> List.copyOf(table.row(attributeScope).values()));
-    } catch (Exception e) {
-      if (LOG_LIMITER.tryAcquire()) {
-        log.error("No attributes available for scope {}", attributeScope);
-      }
-      return Optional.empty();
-    }
+  public List<AttributeMetadata> getAllInScope(
+      @Nonnull RequestContext requestContext, @Nonnull String attributeScope)
+      throws ExecutionException {
+    return getTableForRequestContext(requestContext)
+        .map(table -> List.copyOf(table.row(attributeScope).values()))
+        .orElse(Collections.emptyList());
+  }
+
+  private Optional<Table<String, String, AttributeMetadata>> getTableForRequestContext(
+      RequestContext requestContext) throws ExecutionException {
+    return Optional.of(
+        cache.get(
+            requestContext
+                .getTenantId()
```

Comment:
We want the request context in the key, not using the tenant ID directly. If you don't want to use the context key generated from the request context you can build your own, but either way if I call client.getById and the client needs to make it a network call, it should be using the exact request context I provided to it.